packages/vsx-registry/src/common/vsx-registry-api.ts

```diff
@@ -107,6 +129,56 @@ export class VSXRegistryAPI {
         return fetchText(url);
     }
 
+    /**
+     * Get the latest compatible extension version.
+     * - an extension satisfies compatibility if its `engines.vscode` version is supported.
+     * @param id the extension id.
+     *
+     * @returns the data for the latest compatible extension version if available, else `undefined`.
+     */
+    async getLatestCompatibleExtensionVersion(id: string): Promise<VSXExtensionRaw | undefined> {
+        const extensions = await this.getAllVersions(id);
+
+        for (let i = 0; i < extensions.length; i++) {
+            const extension: VSXExtensionRaw = extensions[i];
+            if (extension.tags?.includes('Built-in')) {
+                return extension;
+            } else if (extension.version === 'latest') {
+                continue;
+            } else if (extension.engines && this.isEngineSupported(extension.engines.vscode)) {
+                return extension;
+            }
+        }
+    }
+
+    /**
+     * Get the latest compatible version of an extension.
+     * @param versions the `allVersions` property.
+     *
+     * @returns the latest compatible version of an extension if it exists, else `undefined`.
+     */
+    getLatestCompatibleVersion(versions: VSXAllVersions[]): VSXAllVersions | undefined {
+        for (const version of versions) {
+            if (this.isEngineSupported(version.engines?.vscode)) {
+                return version;
+            }
+        }
+    }
+
+    /**
+     * Determine if the engine is supported by the application.
+     * @param engine the engine.
+     *
+     * @returns `true` if the engine satisfies the API version.
+     */
+    protected isEngineSupported(engine?: string): boolean {
+        if (!engine) {
+            return false;
+        }
+        const apiVersion = this.apiVersionProvider.getApiVersion();
+        return engine === '*' || semver.satisfies(apiVersion, engine);
```

Comment:
You could make this more efficient by returning `true` if `engine === '*'` before fetching the API version.